UserViewController

barbarakos opened the pull request:
- validating access tokens with check-api
- routing in SceneDelegate

barbarakos committed:
UserViewController

- validating access tokens with check-api
- routing in SceneDelegate

barbarakos committed:
Merge branch 'feature_tokenStorage' into feature_userVC

marinzg commented on QuizApp/AppRouter/AppRouter.swift:
```diff
@@ -9,14 +9,16 @@ class AppRouter: AppRouterProtocol {
+    func showLogIn() {
+        let vc = LoginViewController(router: self)
+        navigationController.setViewControllers([vc], animated: true)
+//        navigationController.pushViewController(vc, animated: false)
```
Why did you leave this commented code?

marinzg commented on QuizApp/CheckClient.swift:
```diff
@@ -0,0 +1,43 @@
+import Foundation
+
+class CheckClient {
```
This logic can go into `LoginClient`.

barbarakos replied:
https://github.com/barbarakos/QuizApp/pull/8

marinzg commented on QuizApp/CheckClient.swift:
```diff
@@ -0,0 +1,43 @@
+    let baseURL = "https://five-ios-quiz-app.herokuapp.com/"
+    let checkPath = "api/v1/check"
+
+    let storage = SecureStorage()
```
`SecureStorage` is not a part of the architecture, but it contains some data so it would be best used in use case or data layer. It should not be used in client.

barbarakos replied:
https://github.com/barbarakos/QuizApp/pull/8

marinzg commented on QuizApp/CheckClient.swift:
```diff
@@ -0,0 +1,43 @@
+        try await executeURLRequest(URLrequest: URLrequest)
+    }
+
+    func executeURLRequest(URLrequest: URLRequest) async throws {
+        guard let (_, response) = try? await URLSession.shared.data(for: URLrequest) else {
+            throw RequestError.serverError
+        }
+
+        guard let httpResponse = response as? HTTPURLResponse else {
+            throw RequestError.dataError
+        }
+
+        if !(200...299).contains(httpResponse.statusCode) {
+            switch httpResponse.statusCode {
+            case 400...499:
+                throw RequestError.clientError
+            case 500...599:
+                throw RequestError.serverError
+            default:
+                throw RequestError.unknown
+            }
+        }
+    }
```
Next PR can be refactor to generic class ApiClient.

barbarakos replied:
https://github.com/barbarakos/QuizApp/pull/8

marinzg commented on QuizApp/SceneDelegate.swift:
```diff
@@ -9,8 +9,26 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+
+        Task {
+            do {
+                let checkClient = CheckClient()
```
Client can be referenced only in Data layer.

barbarakos replied:
https://github.com/barbarakos/QuizApp/pull/8

marinzg commented on QuizApp/User/UserViewController.swift:
```diff
@@ -0,0 +1,34 @@
+
+class UserViewController: UIViewController {
+
+    private var router: AppRouterProtocol!
```
Why does VC has reference to router?

marinzg commented on QuizApp/User/UserViewController.swift:
```diff
@@ -0,0 +1,34 @@
+
+    private var router: AppRouterProtocol!
+
+    convenience init(router: AppRouterProtocol) {
```
Why `convenience`?

marinzg commented on QuizApp/SceneDelegate.swift:
```diff
@@ -9,8 +9,26 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+            }
+        }
+
+        window?.rootViewController = navigationController
```
Since this is an async call before, you'll have a black screen. Do you know how to improve this UX?

marinzg commented on QuizApp/SceneDelegate.swift:
```diff
@@ -9,8 +9,26 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+                }
+            } catch {
+                DispatchQueue.main.async {
+                    router.showLogIn()
```
You can either make each function `@MainActor` or set whole AppRouter to be `@MainActor` to avoid using `DispatchQueue.main`

barbarakos replied:
https://github.com/barbarakos/QuizApp/pull/8

marinzg reviewed (changes requested):
Have you ran this code?
* response model doesn't have `accessToken` property
* LoginViewModel doesn't have required init
* `showLogIn` is not working

barbarakos committed:
TabBarController

- tab bar controller with 3 view controllers
- user's username and log out button on user view controller
- refactor to generic class ApiClient for executing URL requests

barbarakos committed:
Merge branch 'feature_tabbar' into feature_userVC

barbarakos committed:
Revert "TabBarController"

barbarakos committed:
Merge pull request #7 from barbarakos/revert-6-feature_tabbar

Revert "TabBarController"

barbarakos committed:
Some requested changes

barbarakos committed:
Merge branch 'develop' into feature_userVC
